### PR TITLE
Feature 3 overlapping shenanigans

### PR DESCRIPTION
**Pull request by Firebrandt**
This is a PR I'm making for the sake of getting at least partial feedback. Recommend reversion protocol be set up.
Adds a complete interactor, and a dialog window that interacts with it, but that does not satisfy the demands of the use case due to incomplete dependency for the moment. (And dire need of restructuring).

Known issues:
- Straight up doesn't connect to the timetable generator, since that doesn't exist yet. Without intermediate data the use case cannot be run properly. There's an ersatz kind of test case, (it at least RUNS, and the algorithm should work assuming the intermediate data from the timetable generator is entered at some point, but this is pretty bad). 
- Architecture is very not clean, but I will work on that! Notable issues include forgetting about input and output boundaries, which I will add in and restructure. Also dependency injection remains to be done, but will be handled.
- Connection to the main menu UI is incomplete.
- Connection to Hans' use case (another intermediate) is incomplete, but this is my fault for the moment.
- File structure is off, this was deferred.
- Tests incomplete, essentially none.
- Quite a few warnings. 

However, the algorithms are at least complete.
Sorry TA Hindawi, this is half baked kibble, but I'd still appreciate any feedback you can offer I guess.

### CODE REVIEW

**Comment by michaelhsj** [comment]
Hey Michael! It looks like you've got some code that can't quite yet compile, so I hesitate to approve a merge until it can do that.

You've already mentioned it, but I figured this is a good place to talk about what sorts of interfaces you'll need going forward. Your TimeTableMatchCalculateSectionHoursInteractor should implement an InputBoundary of some sort, most likely with the method determineBestMatchingTimetable. Your dialog presenter should implement an OutputBoundary, and again the interface should have all functions that your interactor calls of the presenter.

I can only guess what InputDialogMakingInteractor is, but my advice is that unless this is an entirely separate use case, it might be a task that the presenter does instead? If this has to do with alerting the user and presentation logic, it might be better suited in the presenter. And lastly, whatever View you end up implementing can also implement a View Model, which does something similar.

Also, I noticed that as it is, your Controller seems to be taking Timetable entities as an input when ideally it should only be taking inputs from your View. The only part of your use case that should really be holding or passing around entities is your interactor, so I would suggest using either the TimetableModel that's going to be merged into main from branch 9, or making your interactor an observer to keep track of the timetables instead. Also, I would consider wrapping up the arguments you give the interactor as a RequestModel, but that might not be entirely necessary depending on how much smaller your argument list becomes. 

Looking at the TimeTableMatchCalculateSectionHoursInteractor, a part of me wonders if there's a way to extract a method from calculateTimetableOverlap, if only to make the code more readable.

Also, many of your input and output boundary interfaces have the public access modifier, which is unnecessary and raises IntelliJ warnings, so make sure to clean those up before merging.

You've got some good code so far, so keep working at it! Cleaning up the architecture is the hardest part- everything else will come quite easily.

**Review by michaelhsj — COMMENTED** [review]
Responding to the changes made here: https://github.com/CSC207-2022F-UofT/course-project-todo-find-a-team-name/pull/86#issue-1484118368

Okay! So this is overall looking much, much cleaner than before, so I'm loving the architecture. I've left some comments on the code that you can take a look at, but the long and short is as follows:

I would change the overlap package name to something more descriptive and technical. Right now it seems quite loose.

Your OverlapGeneratedTimetableRelayInteractor doesn't seem to implement any input boundaries. I'm assuming it's the case that only other interactors call it, in which case I don't think there's any problem. But if a controller ends up calling it, make sure it implements some InputBoundary.

Your TimeTableMatchInteractor has some good code, but calculateTimetableOverlap has a 5 levels of nesting, which is a little smelly. If we have some time in the future, we should plan on extracting some of the method to make it easier to read and extend. It's a priority, but the lowest on the list.

TimetableMatchInputBoundary could use a couple sentences of documentation.

The tests in TestCalculateSectionHoursInteractor are all commented out right now. Do these tests work?

Please scan over all of your code for any yellow warnings on IntelliJ. There are too many to list, but I've pointed out a few and  they need to be cleared out before we merge to main.

(For the TA: many of the code comments that are mentioned in this review are here: https://github.com/CSC207-2022F-UofT/course-project-todo-find-a-team-name/pull/86#issue-1484118368
 sorry for all the branches!)

File: src/main/java/overlap_crap_fix_locations_later/application_business/OverlapOutputBoundary.java [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+import retrieve_timetable_use_case.application_business.TimetableModel;
+
+import java.util.List;
+
+public interface OverlapOutputBoundary {
+
+    /**
+     * Pass viewModels to the dialog, given a list of TimetableModels
+     **/
+    public void passViewModelsToDialog(List<TimetableModel> timetableModels);
```
Comment (michaelhsj) [inline src/main/java/overlap_crap_fix_locations_later/application_business/OverlapOutputBoundary.java]: The public access modifier here is unnecessary for an interface. Make sure to remove this (and any other instances in your interfaces).

File: src/main/java/overlap_crap_fix_locations_later/application_business/SectionHoursInputBoundary.java [hunk 2]
```diff
@@ -0,0 +1,12 @@
+package overlap_crap_fix_locations_later.application_business;
+
+
+import retrieve_timetable_use_case.application_business.SectionModel;
+
+/** An input boundary that's used for getting the number of hours in a section. **/
+public interface SectionHoursInputBoundary {
+
+    /** calculate the length, in Hours, of a section. **/
+    public Double calculateHoursOfSection(SectionModel section);
```
Comment (michaelhsj) [inline src/main/java/overlap_crap_fix_locations_later/application_business/SectionHoursInputBoundary.java]: Same here. No need for the public modifier here.

File: src/main/java/overlap_crap_fix_locations_later/application_business/TimeTableMatchInteractor.java [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+        // Compare every section of 2 courses...
+        for (CourseModel mainCourse : mainCourses) {
+            for (CourseModel candidateCourse : candidateCourses) {
+                for (SectionModel mainSection : mainCourse.getSections()) {
+                    for (SectionModel candidateSection: candidateCourse.getSections()){
+                        if (mainSection == candidateSection){
+                            // If they're in the same section in the same course, the two people can go together :).
+                            // Hey, what if we used a companion object to not have to make a new one each time? Or just made this into an object.
+
+                            ArrayList<BlockModel> blockModels = new ArrayList<>();
```
Comment (michaelhsj) [inline src/main/java/overlap_crap_fix_locations_later/application_business/TimeTableMatchInteractor.java]: It looks like this variable is completely unused, which raises a warning on IntelliJ.

File: src/main/java/overlap_crap_fix_locations_later/interface_adapters/ModelToOverlapViewModelConverter.java [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+import retrieve_timetable_use_case.application_business.SectionModel;
+import retrieve_timetable_use_case.application_business.TimetableModel;
+
+import java.util.ArrayList;
+
+public class ModelToOverlapViewModelConverter {
+    /**
+     * Convenience method to turn a BlockModel into an OverlapTimetableBlockViewModel.
+     *
+     * @param blockModel
```
Comment (michaelhsj) [inline src/main/java/overlap_crap_fix_locations_later/interface_adapters/ModelToOverlapViewModelConverter.java]: blockModel is missing a description here.

**Comment by Firebrandt** [comment]
Ok! So I made a kind of duplicate pull request, my bad. 
For the changes in this one, see https://github.com/CSC207-2022F-UofT/course-project-todo-find-a-team-name/pull/86#issue-1484118368 - the comment michael H alluded to. 

Made an additional design decision just now to put UI strings in, strings.properties, a central file. Keeping all our UI strings in one place should make the UI portion of the code more maintainable; a central file makes it easier to proofread and keep track of the UI strings in our app, and if we want to re-use a string resource, that becomes easier too (saves us the trouble of retyping the literal).

In the event that we translate this application to another language, having them all in one places makes translation easier too. 

I added a commit implementing Michael H's revisions, and fixing the warnings in my use case's files.

**Review by michaelhsj — APPROVED** [review]
Looks good! Make sure to resolve your merge conflicts before merging. My advice would be to merge main into your branch and resolve it in IntelliJ first, but you can also get it done using Github's merge UI.

File: src/main/java/overlap_crap_fix_locations_later/application_business/OverlapOutputBoundary.java (hunk 1 above)
Comment (Firebrandt) [reply src/main/java/overlap_crap_fix_locations_later/application_business/OverlapOutputBoundary.java]: Killed it!

File: src/main/java/overlap_crap_fix_locations_later/application_business/TimeTableMatchInteractor.java (hunk 3 above)
Comment (Firebrandt) [reply src/main/java/overlap_crap_fix_locations_later/application_business/TimeTableMatchInteractor.java]: No longer use it!

File: src/main/java/overlap_crap_fix_locations_later/application_business/OverlapOutputBoundary.java (hunk 1 above)
Comment (Firebrandt) [reply src/main/java/overlap_crap_fix_locations_later/application_business/OverlapOutputBoundary.java]: Removed the access modifier.

File: src/main/java/overlap_crap_fix_locations_later/interface_adapters/ModelToOverlapViewModelConverter.java (hunk 4 above)
Comment (Firebrandt) [reply src/main/java/overlap_crap_fix_locations_later/interface_adapters/ModelToOverlapViewModelConverter.java]: The new version fixes this :p.

**Pull request merged**